rust/lance/src/dataset/optimize.rs

```diff
@@ -865,26 +773,42 @@ pub async fn commit_compaction(
             old_fragments: task.original_fragments,
             new_fragments: task.new_fragments,
         };
-        row_id_map.extend(task.row_id_map);
+        if needs_remapping {
+            row_id_map.extend(task.row_id_map);
+        }
         rewrite_groups.push(rewrite_group);
     }
 
-    let index_remapper = options.create_remapper(dataset)?;
-    let affected_ids = rewrite_groups
-        .iter()
-        .flat_map(|group| group.old_fragments.iter().map(|frag| frag.id))
-        .collect::<Vec<_>>();
+    let rewritten_indices = if needs_remapping {
+        let index_remapper = options.create_remapper(dataset)?;
+        let affected_ids = rewrite_groups
+            .iter()
+            .flat_map(|group| group.old_fragments.iter().map(|frag| frag.id))
+            .collect::<Vec<_>>();
 
-    let remapped_indices = index_remapper
-        .remap_indices(row_id_map, &affected_ids)
-        .await?;
-    let rewritten_indices = remapped_indices
-        .iter()
-        .map(|rewritten| RewrittenIndex {
-            old_id: rewritten.original,
-            new_id: rewritten.new,
-        })
-        .collect();
+        let remapped_indices = index_remapper
+            .remap_indices(row_id_map, &affected_ids)
+            .await?;
+        remapped_indices
+            .iter()
+            .map(|rewritten| RewrittenIndex {
+                old_id: rewritten.original,
+                new_id: rewritten.new,
+            })
+            .collect()
+    } else {
+        // In absence of remapping, we did not reserve fragment ids as part of
+        // the individual rewrite tasks. This is partly to avoid the commit
+        // conflict bug described in https://github.com/lancedb/lance/issues/2397
+        // and partly because we don't need to reserve fragment ids if we're not
+        // remapping indices.
+        let new_fragments = rewrite_groups
+            .iter_mut()
+            .flat_map(|group| group.new_fragments.iter_mut())
+            .collect::<Vec<_>>();
+        reserve_fragment_ids(dataset, new_fragments.into_iter()).await?;
```

Comment:
Do we even need to call `reserve_fragment_ids` in this case?  For some reason I thought commit would just assign ids if they were left as -1.

Reply:
That's a good point. We don't need to create another transaction for that.

Reply:
Nevermind, turns out this was in fact load bearing. The assigned fragment ids from this step are used to compute the new fragment bitmap for each of the indices.

Reply:
Ah, thanks for the try